Add per client assigned colors to the dashboard graphs

rmars opened the pull request:
I also moved the CombinedClientGraph into RouterClients for easier control of colours.
The colours go as follows:
Each new client of a router gets the next colour off the preferred colour list.  The latency graph for that client then gets the shades of this colour.

Adding new clients is a little messy, because we have to assign a colour to the new client, and both the CombinedClientGraph and the RouterClient need to know which colour that is.  So the parent (RouterClients) calls an updateColors(newClientToColorMapping) with the new mapping of which clients are assigned which colours.  For now, I regenerate the color mapping for /all/ clients when new clients are added, as this was the most straightforward.  Can definitely change it though. 

Both RouterClient and RouterClients are getting long.  In another review I will split out the latency graph from RouterClient, and move RouterClients to its own file.

Questions: should we be getting latency metrics from `metrics.json` rather than from `/metrics`?

Fixes #242
![screen shot 2016-04-07 at 11 03 41 am](https://cloud.githubusercontent.com/assets/549258/14361804/884fbac8-fcb1-11e5-8951-b0976dfafbe4.png)

![screen shot 2016-04-07 at 11 40 28 am](https://cloud.githubusercontent.com/assets/549258/14362707/9b038092-fcb5-11e5-893e-4ef55b946e2c.png)

esbie commented on linkerd/admin/src/main/resources/io/buoyant/linkerd/admin/js/router_clients.js:
```diff
@@ -122,25 +136,57 @@ var RouterClient = (function() {
+
+      var $container = $("<div />").css("border-top", "2px solid " + colorsForClient.neutral);
+      var $metrics = $("<div />").addClass("col-md-6").appendTo($container);
+      var $chart = $("<div />").addClass("chart-container col-md-6").appendTo($container);
```
please add these as part of a template

esbie commented on linkerd/admin/src/main/resources/io/buoyant/linkerd/admin/js/router_clients.js:
```diff
@@ -122,25 +136,57 @@ var RouterClient = (function() {
+
+    return _.reduce(clients, function(clientMapping, client) {
+      clientMapping[client.prefix] = colors[colorIdx++ % colors.length];
+      return clientMapping;
```
I'd prefer clientMapping to be by `client.label` not by prefix

esbie commented on linkerd/admin/src/main/resources/io/buoyant/linkerd/admin/js/router_clients.js:
```diff
@@ -52,15 +65,15 @@ var RouterClient = (function() {
+  function timeseriesParams(name) {
     return {
-      strokeStyle: colorLookup[colorIdx++ % colorLookup.length], //TODO: #242 per client color scheme
+      strokeStyle: chartLegend[name.replace("request_latency_ms.", "")],
```
I prefer that we don't try parsing metrics in multiple places. Consider something like:
`strokeStyle: chartLegend[name.match(Query.clientQuery().build())[1]]`

rmars replied:
I'm getting the metrics here from client.metrics (from routers.js), which just returns "request_latency_ms.max: 0" etc...  So the client name isn't in name in this case.  

I'm not actually sure whether I should be using client.metrics from routers or if I should be getting latency from metrics.json.

esbie replied:
hm ok fair enough, let's revisit later

esbie commented on linkerd/admin/src/main/resources/io/buoyant/linkerd/admin/js/combined_client_graph.js:
```diff
@@ -4,11 +4,12 @@ var CombinedClientGraph = (function() {
+    function timeseriesParams(name) {
       return {
-        strokeStyle: colors[colorIdx++ % colors.length],
+        strokeStyle: clientColors[name.replace("requests", "")].color,
```
I prefer that we don't try parsing metrics in multiple places. Consider something like:
strokeStyle: clientColors[name.match(Query.clientQuery().build())[1]].color

esbie commented on linkerd/admin/src/main/resources/io/buoyant/linkerd/admin/js/router_controller.js:
```diff
@@ -5,63 +5,77 @@ var RouterController = (function () {
+      color = colors[colorParts[0]][colorParts[1]];
+    } else {
+      color = colors[colorParts[0]].neutral;
+    }
```
simpler as:
`var color = _.property(colorName)(color);`
would have to add .neutrals to the baseColorOrder

esbie commented:
coming along really nicely :)

esbie commented:
:star: :stars: thank you for updating!

rmars commented:
I rebased on top of esbie's branch that moves all the resources.
Also fixed bug where client legends were the wrong colour.

![screen shot 2016-04-08 at 4 26 21 pm](https://cloud.githubusercontent.com/assets/549258/14400067/0a789faa-fda8-11e5-8deb-e3b2bca4e76a.png)

olix0r commented:
:star: for what my review is worth, this looks very nice
